### PR TITLE
Implemented EnableStreamAtPTS to support enabling a separate stream w…

### PR DESCRIPTION
**Pull request by ghost**
…ithout seeking all streams (AudioSettings Change)

This PR enables a much faster switch inbetween Audio streams (language / stereo / dolby) for addons wich have unmuxed (separate) streams.

thanks to @fernetmenta for his patience!

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 1]
```diff
@@ -3764,6 +3776,13 @@ bool CVideoPlayer::CloseStream(CCurrentStream& current, bool bWaitForBuffers)
   if(bWaitForBuffers)
     SetCaching(CACHESTATE_DONE);
 
+  if (STREAM_SOURCE_MASK(current.source) == STREAM_SOURCE_DEMUX)
+  {
+    if (!m_pDemuxer)
+      return false;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: I don't think this is the correct location for doing this. The name of the method "CloseStream" is misleading. Actually is closes the stream player not the demux stream. We still want to be able to close the stream player even if there is no demuxer.

File: xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h [hunk 2]
```diff
@@ -74,7 +74,9 @@ class CInputStreamAddon :
   virtual bool OpenDemux() override;
   virtual DemuxPacket* ReadDemux() override;
   virtual CDemuxStream* GetStream(int iStreamId) override;
+  virtual bool SupportsEnableAtPTS(){ return m_canEnableAtPTS; };
   virtual void EnableStream(int iStreamId, bool enable) override;
+  virtual void EnableStreamAtPTS(int iStreamId, uint64_t pts) override;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h]: "pts" implies DVD_TIME_BASE but this is milliseconds, right?

Comment (mapfau) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h]: It is DVD_TIME_BASE (microseconds). Addon devides it /1000 to come to miliseconds

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 1 above)
Comment (mapfau) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: You mean you will still be able to CloseStream without disabling the demuxer?

File: xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h (hunk 2 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h]: You call it with GetTime() that returns milliseconds

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 1 above)
Comment (mapfau) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: From reading the code: There is an OpenStream() some lines above, it calls "EnableStream(true), then CloseStream() should call EnableStream(false) or do I something wrong.

Situation is, that the stream is not disabled when seek is called on AudioChange, this leads to the fact that I have to seek 3 streams: video / audio old and audio new.

Pls. let me know at what place it has to be called

File: xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h (hunk 2 above)
Comment (mapfau) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h]: Ah, thanks, I didn't know that. Then I'll rename it to EnableStreamAtMs? OK?

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: Closing a stream player and disabling a demux steam are two different stories. Check out this sequence: https://github.com/mapfau/xbmc/blob/inputstream.pts/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L1433
It deletes the demuxer and then closes stream players.

File: xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h (hunk 2 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.h]: or feed pts: DVD_MSEC_TO_TIME(GetTime())

**Comment by FernetMenta** [comment]
thanks!
jenkins build this please

**Comment by MartijnKaijser** [comment]
needs a rebase

**Comment by FernetMenta** [comment]
jenkins build this please

**Pull request merged**